File: app/src/main/java/com/synapse/social/studioasinc/adapter/NotificationAdapter.java
package com.synapse.social.studioasinc.adapter;

import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.google.firebase.auth.FirebaseAuth;
import com.synapse.social.studioasinc.ChatActivity;
import com.synapse.social.studioasinc.HomeActivity;
import com.synapse.social.studioasinc.NotificationConfig;
import com.synapse.social.studioasinc.ProfileActivity;
import com.synapse.social.studioasinc.R;
import com.synapse.social.studioasinc.model.Notification;
import java.util.List;

public class NotificationAdapter extends RecyclerView.Adapter<NotificationAdapter.ViewHolder> {

    private Context context;
    private List<Notification> notificationList;

    public NotificationAdapter(Context context, List<Notification> notificationList) {
        this.context = context;
        this.notificationList = notificationList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.notification_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Notification notification = notificationList.get(position);
        holder.notificationMessage.setText(notification.getMessage());
        
        // Handle notification item clicks for in-app navigation
        holder.itemView.setOnClickListener(v -> {
            handleNotificationClick(notification);
        });
    }

    @Override
    public int getItemCount() {
        return notificationList.size();
    }
    
    /**
     * Handles in-app notification clicks by navigating to appropriate activities.
     * This provides the same deep linking functionality as OneSignal notifications.
     */
    private void handleNotificationClick(Notification notification) {
        try {
            Log.d("NotificationAdapter", "Handling notification click for type: " + notification.getType());
            
            String notificationType = notification.getType();
            if (notificationType == null) {
                Log.w("NotificationAdapter", "Notification type is null, opening HomeActivity");
                openHomeActivity();
                return;
            }
            
            switch (notificationType) {
                case NotificationConfig.NOTIFICATION_TYPE_CHAT_MESSAGE:
                    handleChatMessage(notification);
                    break;
                case NotificationConfig.NOTIFICATION_TYPE_NEW_FOLLOWER:
                    handleNewFollower(notification);
                    break;
                case NotificationConfig.NOTIFICATION_TYPE_PROFILE_LIKE:
                    handleProfileLike(notification);
                    break;
                default:
                    // Handle post-related notifications (likes, comments, replies, new posts, mentions)
                    if (notificationType.equals(NotificationConfig.NOTIFICATION_TYPE_NEW_LIKE_POST) ||
                        notificationType.equals(NotificationConfig.NOTIFICATION_TYPE_NEW_COMMENT) ||
                        notificationType.equals(NotificationConfig.NOTIFICATION_TYPE_NEW_REPLY) ||
                        notificationType.equals(NotificationConfig.NOTIFICATION_TYPE_NEW_POST) ||
                        notificationType.equals(NotificationConfig.NOTIFICATION_TYPE_MENTION_POST)) {
                        handlePostNotification(notification);
                    } else {
                        Log.w("NotificationAdapter", "Unknown notification type: " + notificationType + ", opening HomeActivity");
                        openHomeActivity();
                    }
                    break;
            }
            
        } catch (Exception e) {
            Log.e("NotificationAdapter", "Error handling notification click", e);
            openHomeActivity();
        }
    }
    
    private void handleChatMessage(Notification notification) {
        try {
            String senderUid = notification.getFrom();
            if (senderUid == null || senderUid.isEmpty()) {
                Log.e("NotificationAdapter", "Chat notification missing sender UID");
                openHomeActivity();
                return;
            }
            
            Intent intent = new Intent(context, ChatActivity.class);
            intent.putExtra("uid", senderUid);
            intent.putExtra("origin", "NotificationClick");
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            context.startActivity(intent);
            
        } catch (Exception e) {
            Log.e("NotificationAdapter", "Error handling chat message notification", e);
            openHomeActivity();
        }
    }
    
    private void handleNewFollower(Notification notification) {
        try {
            String senderUid = notification.getFrom();
            if (senderUid == null || senderUid.isEmpty()) {
                Log.e("NotificationAdapter", "New follower notification missing sender UID");
                openHomeActivity();
                return;
            }
            
            Intent intent = new Intent(context, ProfileActivity.class);
            intent.putExtra("uid", senderUid);
            intent.putExtra("origin", "NotificationClick");
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            context.startActivity(intent);
            
        } catch (Exception e) {
            Log.e("NotificationAdapter", "Error handling new follower notification", e);
            openHomeActivity();
        }
    }
    
    private void handleProfileLike(Notification notification) {
        try {
            String senderUid = notification.getFrom();
            if (senderUid == null || senderUid.isEmpty()) {
                Log.e("NotificationAdapter", "Profile like notification missing sender UID");
                openHomeActivity();
                return;
            }
            
            Intent intent = new Intent(context, ProfileActivity.class);
            intent.putExtra("uid", senderUid);
            intent.putExtra("origin", "NotificationClick");
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            context.startActivity(intent);
            
        } catch (Exception e) {
            Log.e("NotificationAdapter", "Error handling profile like notification", e);
            openHomeActivity();
        }
    }
    
    private void handlePostNotification(Notification notification) {
        try {
            String postId = notification.getPostId();
            String senderUid = notification.getFrom();
            
            Intent intent = new Intent(context, HomeActivity.class);
            if (postId != null && !postId.isEmpty()) {
                intent.putExtra("postId", postId);
            }
            if (senderUid != null && !senderUid.isEmpty()) {
                intent.putExtra("senderUid", senderUid);
            }
            intent.putExtra("origin", "NotificationClick");
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            context.startActivity(intent);
            
        } catch (Exception e) {
            Log.e("NotificationAdapter", "Error handling post notification", e);
            openHomeActivity();
        }
    }
    
    private void openHomeActivity() {
        try {
            Intent intent = new Intent(context, HomeActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            context.startActivity(intent);
        } catch (Exception e) {
            Log.e("NotificationAdapter", "Failed to open HomeActivity as fallback", e);
        }
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public TextView notificationMessage;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            notificationMessage = itemView.findViewById(R.id.notification_message);
        }
    }
}

File: app/src/main/java/com/synapse/social/studioasinc/adapter/NotificationAdapter.kt
package com.synapse.social.studioasinc.adapter

import android.content.Context
import android.content.Intent
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.synapse.social.studioasinc.ProfileActivity
import com.synapse.social.studioasinc.R
import com.synapse.social.studioasinc.model.Notification
import com.synapse.social.studioasinc.model.User
import de.hdodenhof.circleimageview.CircleImageView

class NotificationAdapter(
    private val context: Context,
    private val notificationList: List<Notification>
) : RecyclerView.Adapter<NotificationAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.notification_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val notification = notificationList[position]

        holder.notificationMessage.text = notification.message
        holder.timestamp.text = getRelativeTime(notification.timestamp)
        holder.unreadIndicator.visibility = if (notification.isRead) View.GONE else View.VISIBLE

        setNotificationIcon(holder.notificationTypeIcon, notification.type)
        loadUserAvatar(holder.userAvatar, notification.from)

        holder.itemView.setOnClickListener {
            if ("follow" == notification.type) {
                val intent = Intent(context, ProfileActivity::class.java).apply {
                    putExtra("uid", notification.from)
                }
                context.startActivity(intent)
            } else {
                // Assuming other notifications lead to a post
                // val intent = Intent(context, PostActivity::class.java).apply {
                //     putExtra("postId", notification.postId)
                // }
                // context.startActivity(intent)
            }
        }
    }

    override fun getItemCount(): Int = notificationList.size

    private fun setNotificationIcon(imageView: ImageView, type: String?) {
        val iconRes = when (type) {
            "like" -> R.drawable.ic_favorite_48px
            "comment" -> R.drawable.ic_comment_48px
            "follow" -> R.drawable.ic_person_add_48px
            else -> R.drawable.ic_notifications_48px // A default icon
        }
        imageView.setImageResource(iconRes)
    }

    private fun loadUserAvatar(imageView: CircleImageView, userId: String) {
        FirebaseDatabase.getInstance().getReference("users").child(userId)
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(snapshot: DataSnapshot) {
                    val user = snapshot.getValue(User::class.java)
                    if (user?.profileImageUrl?.isNotEmpty() == true) {
                        Glide.with(context).load(user.profileImageUrl).into(imageView)
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                    // Handle error
                }
            })
    }

    private fun getRelativeTime(time: Long): String {
        return DateUtils.getRelativeTimeSpanString(
            time,
            System.currentTimeMillis(),
            DateUtils.MINUTE_IN_MILLIS
        ).toString()
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val userAvatar: CircleImageView = itemView.findViewById(R.id.user_avatar)
        val notificationTypeIcon: ImageView = itemView.findViewById(R.id.notification_type_icon)
        val notificationMessage: TextView = itemView.findViewById(R.id.notification_message)
        val timestamp: TextView = itemView.findViewById(R.id.timestamp)
        val unreadIndicator: View = itemView.findViewById(R.id.unread_indicator)
    }
}
